runtime/runtime-core/jvm/src/aws/smithy/kotlin/runtime/io/GzipByteReadChannel.kt

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.io
+
+import aws.smithy.kotlin.runtime.InternalApi
+import java.util.zip.GZIPOutputStream
+
+@InternalApi
+public class GzipByteReadChannel(
+    private val channel: SdkByteReadChannel,
+) : SdkByteReadChannel by channel {
+    private val gzipBuffer = SdkBuffer()
+    private val gzipOutputStream = GZIPOutputStream(gzipBuffer.outputStream())
+
+    /**
+     * Keeps track of whether a read operation has been made on this byte read channel
+     */
+    private var read: Boolean = false
```

Comment:
question: This seems suspicious, why is it necessary?

Reply:
When compressing an empty payload the `GzipByteReadChannel` sets `isClosedForRead` to true from the first read. The program will go into that first `if` block checking the value of `isClosedForRead` and would return -1 and do nothing with the `gzipOutputStream`. The reason it's needed is to write an empty byte array into the `gzipOutputStream` without that it can't be decompressed. A byte array that looks like "[]" can't be decompressed by gzip, it needs the header. It also closes the gzip output steam and buffer since they won't be needed anymore

Reply:
I'm not convinced we need this. If we're "compressing" an empty payload then it should be a no-op and we shouldn't compress anything or set a `Content-Encoding`.